packages/agent/src/certificate.ts

```diff
@@ -130,28 +129,41 @@ export class Certificate {
     }
   }
 
-  private async _checkDelegation(d?: Delegation): Promise<ArrayBuffer> {
+  private async _checkDelegationAndGetKey(
+    canisterId: Principal,
+    d?: Delegation,
+  ): Promise<ArrayBuffer> {
     if (!d) {
-      if (!this._rootKey) {
-        if (this._agent.rootKey) {
-          this._rootKey = this._agent.rootKey;
-          return this._rootKey;
-        }
-
-        throw new Error(`Agent does not have a rootKey. Do you need to call 'fetchRootKey'?`);
-      }
       return this._rootKey;
     }
-    const cert: Certificate = new Certificate(d as any, this._agent);
-    if (!(await cert.verify())) {
+    const cert: Certificate = new Certificate(d.certificate, this._rootKey);
+    if (!(await cert.verify(canisterId))) {
       throw new Error('fail to verify delegation certificate');
     }
 
-    const lookup = cert.lookup(['subnet', d.subnet_id, 'public_key']);
-    if (!lookup) {
+    if (canisterId.compareTo(Principal.managementCanister()) != 'eq') {
+      const range_lookup = cert.lookup(['subnet', d.subnet_id, 'canister_ranges']);
```

Comment:
Thanks for catching this - I changed all the `snake_case` I introduced to `camelCase`. However, `subnet_id` was there before, so I didn't touch this - I'm not sure if there's some magic going on because it's parsed from a `subnet_id` field in CBOR.